feat: revise task interpreter with process execution

phreed opened the pull request:
Replaces https://github.com/fecet/pixi/pull/1

This commit consolidates interpreter functionality by:
- Allow interpreter to be string or array of strings for flexibility in selecting interpreters
- Use Rust's std::process::Command instead of deno_task_shell for interpreter tasks
- Add comprehensive test coverage for various interpreter scenarios
- Support passing flags and arguments to interpreters (e.g., ['python', '-u'])
- Maintain backward compatibility with existing task definitions
- Add schema validation for interpreter field formats
- Include test projects demonstrating interpreter usage patterns

fecet committed:
feat: add custom interpreter support for pixi tasks

fecet committed:
feat: allow stdin in pixi task

phreed commented:
At some point nushell `nu` was modified to not read its script from stdin.

I have chatted with them about this and they make some good points.
https://www.explainxkcd.com/wiki/index.php/341:_1337:_Part_1
How do you know when the script is complete?

fecet commented:
Thanks for your help! I want to know if creating temporary files is necessary for Nushell or your scenario. As far as I know, nushell can also use --stdin to allow accepting standard input?

The approach of using temporary files brings some complexity, and dealing with file extensions is also rather tricky.

fecet committed:
refactor!: use tmp file instead of stdin

phreed commented:
> Thanks for your help! I want to know if creating temporary files is necessary for Nushell or your scenario. As far as I know, nushell can also use --stdin to allow accepting standard input?

That is what I thought as well but it turns out to not be the case.

In the past nushell would receive scripts on stdin but that was removed (I am not sure which version).
At present there are several ways to get nushell to process a script:

1. Via a named file `nu foo.nu`
2. Via a command line argument `nu -c 'print "hi"'`
3. Via the interpreter `nu` which processes commands on stdin if a tty : so not really a script
4. Combined command line and interpreter `nu -e 'print "hi"'` 

The `--stdin` causes stdin to be directed to the script (or -c command) not to be interpreted as a script.
The behavior of stdin when none of these cases is in play is unclear to me.

> The approach of using temporary files brings some complexity, and dealing with file extensions is also rather tricky.

Yes, it adds complexity in order to remove ambiguity.
I think the main issue they are trying to deal with is the ambiguity of when the stdin script is complete.
https://www.explainxkcd.com/wiki/index.php/341:_1337:_Part_1

phreed commented:
I agree, the file extension bit is a bad idea and should be retracted.

phreed commented:
I have been looking for the discussion on the subject of piping scripts as I think it is relevant to all interpreters.

The idea of Nushell directly executing a script from stdin (e.g., echo "print 'hi'" | nu) has been discussed and generally rejected or deprioritized by the core developers.

The primary reason for this rejection often comes down to:

### Complexity and Ambiguity: 
Nushell's design emphasizes structured data and pipelines. Allowing direct script execution from stdin could introduce ambiguity about whether the input is data to be processed by a command, or code to be executed. The nu -c "command" flag is the designated way to run a command string directly.
https://www.nushell.sh/book/how_nushell_code_gets_run.html
"Think of Nushell as a compiled language" 

### Security Concerns: 
Executing arbitrary code from an untrusted stdin source can be a security risk. While this can be mitigated, it adds complexity to the interpreter.

### Prioritization: 
The Nushell team has a lot of features and improvements to work on. Implementing stdin script execution might be seen as less critical compared to other core features, bug fixes, or performance improvements, especially when there are existing workarounds (like the temp file method or nu -c).

### Philosophical Alignment: 
Nushell aims to be a modern shell with a clear and consistent mental model. The -c flag for inline command execution and file arguments for script execution fit this model. Piping to nu is generally expected to pipe data into a command, not the script itself.

phreed committed:
feat: implement basic interpreter support with array format

- Change interpreter from Option<String> to Option<Vec<String>>
- Add TOML parsing support for both string and array formats
- Update schema to support both string and array interpreter formats
- Add execute_with_interpreter method for std::process::Command execution
- Add basic interpreter examples to full.toml
- Handle interpreter execution via stdin pipe approach

phreed committed:
feat: implement enhanced interpreter execution with placeholder support

- Add support for '-' placeholder (temporary file approach)
- Add support for '@' placeholder (inline script approach)
- Add support for '<' placeholder (explicit stdin, discouraged)
- Add warning messages for discouraged approaches
- Add automatic file extension detection (.py, .nu, .sh, etc.)
- Add comprehensive nushell.toml example file
- Add interpreter_tempfile_tests integration tests
- Support multiple placeholders in same command
- Maintain backward compatibility with traditional stdin approach

This enables reliable execution of interpreters like nushell that don't
work well with stdin-based script execution.

phreed committed:
fix: ensure interpreter tasks work with execute_with_pipes method

- Add interpreter handling to execute_with_pipes method for test compatibility
- Fix test builder to properly parse interpreter strings into argument arrays
- All interpreter tests now passing

This ensures that both the CLI path and test framework can execute
interpreter tasks correctly.

phreed committed:
Enhance interpreter support with string format option

Previously, interpreters were only supported in array format. This
change adds support for string format interpreters with a {0}
placeholder, providing a simpler and more flexible syntax for common use
cases.

The interpreter field now accepts both formats:

```toml
interpreter = "python {0}" # Explicit placeholder interpreter = "bash" #
Implicit append

interpreter = ["python", "-"] # File mode interpreter = ["python", "-c",
"@"] # Inline mode ```

String format automatically handles file extensions and supports
interpreter flags in a more natural way. Array format remains for
advanced use cases.

phreed committed:
Remove file extension detection and preserve original schema formatting
